components/hab/src/command/plan/render.rs

```diff
@@ -0,0 +1,176 @@
+    // for now assume we're working in the plan dir if --default-toml not passed
+    let default_toml = read_to_string(&default_toml_path)
+        .expect(&format!("Something went wrong reading: {}", &default_toml_path));
+    // merge default into data struct
+    merge(&mut data, toml_to_json(&default_toml));
+
+    // import default.toml values, convert to JSON
+    // ui.begin(format!("Importing user.toml: {}", &user_toml_path));
+    let user_toml = match user_toml_path {
+        Some(path) => {
+            if !(quiet) {
+              // print helper message, maybe only print if '--verbose'? how?
+              ui.begin(format!("Importing user.toml: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+                .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        None => "".to_string(),
+    };
+    // merge default into data struct
+    merge(&mut data, toml_to_json(&user_toml));
+
+    // read mock data if provided
+    let mock_data = match mock_data_path {
+        Some(path) => {
+            if !(quiet) {
+                // print helper message, maybe only print if '--verbose'? how?
+                ui.begin(format!("Importing override file: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+              .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        // return an empty json block if '--mock-data' isn't defined.
+        // this allows us to merge an empty JSON block
+        None => "{}".to_string(),
+    };
+    // merge mock data into data
+    merge(&mut data, serde_json::from_str(&mock_data).unwrap());
+
+    // create a template renderer
+    let mut renderer = TemplateRenderer::new();
+    // register our template 
+    renderer
+        .register_template_string(&template_path, &template)
+        .expect("Could not register template content");
+    // render our JSON override in our template.
+    let rendered_template = renderer.render(&template_path, &data).ok().unwrap();
+    
+    if print {
+        if !(quiet) {
+          ui.br()?;
+          ui.warn(format!("###======== Rendered template: {}", &template_path))?;
+        }
+
+        println!("{}", rendered_template);
+
+        if !(quiet) {
+          ui.warn(format!("========### End rendered template: {}", &template_path))?;
+        }
+    }
+
+    // if not no render dir (aka "unless no_render_dir == true")
+    if !(no_render_dir) {
+      // Render our template file
+      create_with_template(ui, &format!("{}/{}", render_dir, file_name), &rendered_template, quiet)?;
+    }
+
+    if !(quiet) {
+      ui.br()?;
+    }
+    // not really sure this is correct...
+    Ok(())
+}
+
+fn toml_to_json(cfg: &str) -> Json {
+    // parse TOML string to Value
+    let toml_value = cfg.parse::<Value>().expect("Error parsing TOML");
```

Comment:
When I commit this suggestion, I get:

```
133 |     let toml_value = cfg.parse::<Value>()?;
    |                      ^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `serde_json::Value`
```

Which I think makes sense.  And why I have to change the return type of the function to `Result<Json>`.

When I do that I get a:

```
error[E0277]: the trait bound `error::Error: config::_IMPL_SERIALIZE_FOR_Config::_serde::Deserialize<'_>` is not satisfied
   --> components/hab/src/command/plan/render.rs:137:5
    |
137 |     serde_json::from_str(&format!(r#"{{ "cfg": {} }}"#, &toml_string)).unwrap()
    |     ^^^^^^^^^^^^^^^^^^^^ the trait `config::_IMPL_SERIALIZE_FOR_Config::_serde::Deserialize<'_>` is not implemented for `error::Error`
```

Which I don't really understand, but I'm fairly certain is caused by the `.unwrap()`.

Removing the `.unwrap()` yields a:

```
error[E0308]: mismatched types
   --> components/hab/src/command/plan/render.rs:137:5
    |
131 | fn toml_to_json(cfg: &str) -> Result<Json> {
    |                               ------------ expected `std::result::Result<serde_json::Value, error::Error>` because of return type
...
137 |     serde_json::from_str(&format!(r#"{{ "cfg": {} }}"#, &toml_string))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `error::Error`, found struct `serde_json::Error`
```

Which is why I think you pointed at `hab/src/error.rs`.

But... I'm kinda stumped...  I [added an error type](https://github.com/qubitrenegade/habitat-1/commit/861451534ad62bf24087c6746f465e2f2c1c52b4) (I think), but this still yields the same error message...

Reply:
Yeah, this one is a bit subtle. For the approach that seems best to me, you need to change three things:

1. The `toml_to_json` function:
    ```rust
    fn toml_to_json(cfg: &str) -> Result<Json> {
        let toml_value = cfg.parse::<Value>()?;
        let toml_string = serde_json::to_string(&toml_value)?;
        let json = serde_json::from_str(&format!(r#"{{ "cfg": {} }}"#, &toml_string))?;
        Ok(json)
    }
    ```
    We could just return `serde_json::from_str(…)`, but then our function would have to return `std::result::Result<Json, serde_json::Error>>` instead of [our crate-local `Result`](https://github.com/habitat-sh/habitat/blob/master/components/hab/src/error.rs#L33) type. Instead, I added an extra line with `?` to do an implied conversion and return `Ok(json)` on the next line.
1. Update the callers of `toml_to_json` with a `?` on the end like:
    ```rust
    merge(&mut data, toml_to_json(&default_toml)?);
    ```
1. Add a new error variant and implement `From<serde_json::Error>` for it.

Reply:
I think I understand why we use it here, but I'm maybe not clear on all the places/reasons we use `Ok()`.  What I'm not clear on is the connection between the `Result` defined in `error.rs`.

That is to say, I successfully added the error type, it gives a nice error:

```
» Importing override file: "./test/fixtures/render/error/override.json"
✗✗✗
✗✗✗ expected `,` or `}` at line 4 column 5
✗✗✗
```

But I'd be hard pressed to explain _how_ it works :slightly_smiling_face:

Reply:
Have you read https://blog.burntsushi.net/rust-error-handling ? It's an excellent primer on rust's approach to error-handling. It can certainly be confusing at first since it's pretty different from most imperative and OO languages, but it's really elegant once you get your head around it.

If you have any questions after reading, I'd be happy to discuss.